src/Microsoft.IdentityModel.Tokens/ProvidersCache/HashCache.cs

```diff
@@ -0,0 +1,264 @@
+        {
+            value = default;
+
+            var found = _map.TryRemove(key, out ProviderCacheItem<TKey, TValue> val);
+            if (found)
+                value = val;
+
+            return found;
+        }
+
+        /// <summary>
+        /// Validate the key, value and check the expiration time before adding the item is added to the cache.
+        /// </summary>
+        /// <param name="key">The key of the item to be added to the cache.</param>
+        /// <param name="value">The value of the item to be added to the cache.</param>
+        /// <param name="expirationTime">The expiration time of the item.</param>
+        /// <returns>true if valid, false otherwise.</returns>
+        protected static bool ValidateValues(TKey key, TValue value, DateTime expirationTime)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (value == null)
+                throw LogHelper.LogArgumentNullException(nameof(value));
+
+            // if item already expired, do not add it to the cache
+            return expirationTime >= DateTime.UtcNow;
+        }
+
+        /// <summary>
+        /// When the cache is at _maxCapacityPercentage, it needs to be compacted by _compactionPercentage.
+        /// This method calculates the new size of the cache after being compacted.
+        /// </summary>
+        /// <returns>The new target cache size after compaction.</returns>
+        protected int CalculateNewCacheSize()
+        {
+            // use the smaller of _map.Count and _capacity
+            int currentCount = Math.Min(_map.Count, _capacity);
+            return currentCount - (int)(currentCount * _compactionPercentage);
+        }
+
+        /// <summary>
+        /// Determine if the cache needs compaction.
+        /// </summary>
+        protected bool NeedsCompaction => (double)_map.Count / _capacity >= _maxCapacityPercentage;
+
+        #endregion
+
+        /// <summary>
+        /// Remove items from the map by the desired compaction percentage.
+        /// Since this is a simple hash-based cache that does not track the last used time of items, simply remove the desired number of items.
+        /// This should be a private method.
+        /// </summary>
+        private void CompactCache(object state)
+        {
+            // use the _capacity for the newCacheSize calculation in the case where the cache is experiencing overflow
+            var newCacheSize = CalculateNewCacheSize();
+            while (_map.Count > newCacheSize)
+            {
+                var item = _map.FirstOrDefault();
+                if (_map.TryRemove(item.Key, out var cacheItem))
+                    OnItemRemoved?.Invoke(cacheItem.Value);
+            }
+
+            // reset the _compactionState to CompactionStateNotRunning
+            _compactionState = CompactionStateNotRunning;
```

Comment:
>_compactionState = CompactionStateNotRunning; [](http://example.com/codeflow?start=12&length=45)

seems like this should be set to running when starting the compacting operation

Reply:
Yes, doesn't seem to be set to running.

Reply:
The compaction is started i the SetValue() method when the cache exceeds the limit (_compactionState will be set to CompactionStateInProgress).
This is at the end of the compacting operation so the state should be set to CompactionStateNotRunning.